package/MDAnalysis/topology/MMCIFParser.py

```diff
@@ -0,0 +1,178 @@
+"""
+MMCIF Topology Parser # 
+===================
+"""
+
+import gemmi
+import numpy as np
+import warnings
+import itertools
+
+from ..core.topology import Topology
+from ..core.topologyattrs import (
+    AltLocs,
+    Atomids,
+    Atomnames,
+    Atomtypes,
+    ChainIDs,
+    Elements,
+    FormalCharges,
+    ICodes,
+    Masses,
+    Occupancies,
+    RecordTypes,
+    Resids,
+    Resnames,
+    Resnums,
+    Segids,
+    Tempfactors,
+)
+from .base import TopologyReaderBase
+
+
+def _into_idx(arr: list[int]) -> list[int]:
+    return [idx for idx, (_, group) in enumerate(itertools.groupby(arr)) for _ in group]
+
+
+class MMCIFParser(TopologyReaderBase):
+    format = "MMCIF"
+
+    def parse(self, **kwargs):
+        """Read the file and return the structure.
+
+        Returns
+        -------
+        MDAnalysis Topology object
+        """
+        structure = gemmi.read_structure(self.filename)
+
+        if len(structure) > 1:
+            warnings.warn(
+                "MMCIF model {self.filename} contains {len(model)=} different models, "
+                "but only the first one will be used to assign the topology"
+            )
+        model = structure[0]
+
+        # atom properties
+        (
+            altlocs,  # at.altloc
+            serials,  # at.serial
+            names,  # at.name
+            atomtypes,  # at.name
+            # ------------------
+            chainids,  # chain.name
+            elements,  # at.element.name
+            formalcharges,  # at.charge
+            weights,  # at.element.weight
+            # ------------------
+            occupancies,  # at.occ
+            record_types,  # res.het_flag
+            tempfactors,  # at.b_iso
+            residx,  # _into_idx(res.seqid.num)
+        ) = map(  # this is probably not pretty, but it's efficient -- one loop over the mmcif
```

Comment:
Are all the fields here guaranteed in a valid pdbx? One benefit to working column by column is that you can do optional columns

Reply:
Do you have an example of a PDBx in mind, or like a test set for them? I've never actually worked with the format, since in RCSB afaik we have only pdb or mmcif

Reply:
PDBx is mmcif.  The download links here will give you an example file: https://www.rcsb.org/structure/4ake we use 4ake elsewhere in the testsuite.  In my experience, sometimes the PDB / mmcif versions of the same entry aren't completely identical, so I wouldn't worry about trying to align the PDB & PDBx tests.